src/Form/DeliveryReferencedContentConfirmForm.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace Drupal\delivery\Form;
+
+use Drupal\Core\Form\ConfirmFormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
+use Drupal\delivery\DeliveryCartReferencedContent;
+use Drupal\delivery\DeliveryCartService;
+use Drupal\workspaces\Entity\Workspace;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class DeliveryReferencedContentConfirmForm extends ConfirmFormBase {
+
+  protected $deliveryCart;
+
+  protected $cart = [];
+
+  protected $cartEntities = [];
+
+  /**
+   * DeliveryEmptyCartConfirmForm constructor.
+   *
+   * @param DeliveryCartService $delivery_cart
+   */
+  public function __construct(DeliveryCartService $delivery_cart) {
+    $type = 'node';
+    $this->deliveryCart = $delivery_cart;
+    if($cartItems = $this->deliveryCart->getCart()) {
+      $entityIds = $cartItems[$type] ?? [];
+      $entityStorage = \Drupal::entityTypeManager()->getStorage($type);
+      foreach ($entityIds as $entityIdData) {
+        $sourceWorkspace = Workspace::load($entityIdData['workspace_id']);
+        $entity = $entityStorage->loadRevision($entityIdData['revision_id']);
```

Comment:
In case we have a lot of items in the cart (like thousands), you will get an out of memory error here. I would say that parts of this code should be moved into the getDescription(), because there you need the entity label, so you load only 100 entities max. Then in the submit handler you can actually directly access the cart, and get the entity id and revision id from there and pass them as parameters to the batch operation. Then, in the batch operation itself (so findReferencedContent()) you can load the entity.